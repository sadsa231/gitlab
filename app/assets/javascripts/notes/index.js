import Vue from 'vue';
import { parseBoolean } from '~/lib/utils/common_utils';
import { getLocationHash } from '~/lib/utils/url_utility';
import NotesApp from './components/notes_app.vue';
import { store } from './stores';
import { getNotesFilterData } from './utils/get_notes_filter_data';

export default () => {
  const el = document.getElementById('js-vue-notes');
  if (!el) {
    return;
  }

  const notesFilterProps = getNotesFilterData(el);
  const showTimelineViewToggle = parseBoolean(el.dataset.showTimelineViewToggle);

  const notesDataset = el.dataset;
  const parsedUserData = JSON.parse(notesDataset.currentUserData);
  const noteableData = JSON.parse(notesDataset.noteableData);
  let currentUserData = {};

  noteableData.noteableType = notesDataset.noteableType;
  noteableData.targetType = notesDataset.targetType;
  noteableData.discussion_locked = parseBoolean(noteableData.discussion_locked);

  if (parsedUserData) {
    currentUserData = {
      id: parsedUserData.id,
      name: parsedUserData.name,
      username: parsedUserData.username,
      avatar_url: parsedUserData.avatar_path || parsedUserData.avatar_url,
      path: parsedUserData.path,
      can_add_timeline_events: parseBoolean(notesDataset.canAddTimelineEvents),
    };
  }

  const notesData = JSON.parse(notesDataset.notesData);

  store.dispatch('setNotesData', notesData);
  store.dispatch('setNoteableData', noteableData);
  store.dispatch('setUserData', currentUserData);
  store.dispatch('setTargetNoteHash', getLocationHash());
  store.dispatch('fetchNotes');

  // eslint-disable-next-line no-new
  new Vue({
    el,
    name: 'NotesRoot',
    components: {
      NotesApp,
    },
    store,
    provide: {
      showTimelineViewToggle,
      reportAbusePath: notesDataset.reportAbusePath,
    },
    data() {
      return {
        noteableData,
        currentUserData,
        notesData: JSON.parse(notesDataset.notesData),
      };
    },
    render(createElement) {
      return createElement('notes-app', {
        props: {
          noteableData: this.noteableData,
          notesData: this.notesData,
          userData: this.currentUserData,
          ...notesFilterProps,
        },
      });
    },
  });
};
