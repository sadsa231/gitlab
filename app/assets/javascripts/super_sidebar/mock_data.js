import { s__ } from '~/locale';

export const context = {
  title: 'Typeahead.js',
  link: '/',
  avatar: 'https://gitlab.com/uploads/-/system/project/avatar/278964/project_avatar.png?width=32',
};

export const contextSwitcherItems = {
  yourWork: { title: s__('Navigation|Your work'), link: '/', icon: 'work' },
  recentProjects: [
    {
      // eslint-disable-next-line @gitlab/require-i18n-strings
      title: 'Orange',
      subtitle: 'tropical-tree',
      link: '/tropical-tree',
      avatar:
        'https://gitlab.com/uploads/-/system/project/avatar/4456656/pajamas-logo.png?width=64',
    },
    {
      // eslint-disable-next-line @gitlab/require-i18n-strings
      title: 'Lemon',
      subtitle: 'tropical-tree',
      link: '/tropical-tree',
      avatar: 'https://gitlab.com/uploads/-/system/project/avatar/7071551/GitLab_UI.png?width=64',
    },
    {
      // eslint-disable-next-line @gitlab/require-i18n-strings
      title: 'Coconut',
      subtitle: 'tropical-tree',
      link: '/tropical-tree',
      avatar:
        'https://gitlab.com/uploads/-/system/project/avatar/4149988/SVGs_project.png?width=64',
    },
  ],
  recentGroups: [
    {
      title: 'Developer Evangelism at GitLab',
      subtitle: 'tropical-tree',
      link: '/tropical-tree',
      avatar:
        'https://gitlab.com/uploads/-/system/group/avatar/10087220/rainbow_tanuki.jpg?width=64',
    },
    {
      title: 'security-products',
      subtitle: 'tropical-tree',
      link: '/tropical-tree',
      avatar:
        'https://gitlab.com/uploads/-/system/group/avatar/11932235/gitlab-icon-rgb.png?width=64',
    },
    {
      title: 'Tanuki-Workshops',
      subtitle: 'tropical-tree',
      link: '/tropical-tree',
      avatar:
        'https://gitlab.com/uploads/-/system/group/avatar/5085244/Screenshot_2019-04-29_at_16.13.07.png?width=64',
    },
  ],
};
