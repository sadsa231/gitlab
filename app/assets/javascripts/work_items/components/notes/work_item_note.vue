<script>
import { GlAvatarLink, GlAvatar } from '@gitlab/ui';
import TimelineEntryItem from '~/vue_shared/components/notes/timeline_entry_item.vue';
import NoteBody from '~/work_items/components/notes/work_item_note_body.vue';
import NoteHeader from '~/notes/components/note_header.vue';
import NoteActions from '~/work_items/components/notes/work_item_note_actions.vue';
import { renderGFM } from '~/behaviors/markdown/render_gfm';

export default {
  name: 'WorkItemNoteThread',
  components: {
    TimelineEntryItem,
    NoteBody,
    NoteHeader,
    NoteActions,
    GlAvatar,
    GlAvatarLink,
  },
  props: {
    note: {
      type: Object,
      required: true,
    },
    isFirstNote: {
      type: Boolean,
      required: false,
      default: false,
    },
  },
  computed: {
    author() {
      return this.note.author;
    },
    entryClass() {
      return {
        'note note-wrapper note-comment': true,
        'gl-p-4': !this.isFirstNote,
      };
    },
    showReply() {
      return this.note.userPermissions.createNote && this.isFirstNote;
    },
  },
  methods: {
    renderGFM() {
      renderGFM(this.$refs['note-body']);
    },
    showReplyForm() {
      this.$emit('startReplying');
    },
  },
};
</script>

<template>
  <timeline-entry-item :class="entryClass">
    <div v-if="!isFirstNote" :key="note.id" class="timeline-avatar gl-float-left">
      <gl-avatar-link :href="author.webUrl">
        <gl-avatar
          :src="author.avatarUrl"
          :entity-name="author.username"
          :alt="author.name"
          :size="32"
        />
      </gl-avatar-link>
    </div>
    <div class="timeline-content-inner">
      <div class="note-header">
        <note-header :author="author" :created-at="note.createdAt" :note-id="note.id" />
        <note-actions :show-reply="showReply" @startReplying="showReplyForm" />
      </div>
      <div class="timeline-discussion-body">
        <note-body ref="noteBody" :note="note" />
      </div>
    </div>
  </timeline-entry-item>
</template>
