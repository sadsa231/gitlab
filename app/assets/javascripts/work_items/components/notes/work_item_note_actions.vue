<script>
import ReplyButton from '~/notes/components/note_actions/reply_button.vue';

export default {
  name: 'WorkItemNoteActions',
  components: {
    ReplyButton,
  },
  props: {
    showReply: {
      type: Boolean,
      required: true,
    },
  },
};
</script>

<template>
  <div class="note-actions">
    <reply-button v-if="showReply" ref="replyButton" @startReplying="$emit('startReplying')" />
  </div>
</template>
