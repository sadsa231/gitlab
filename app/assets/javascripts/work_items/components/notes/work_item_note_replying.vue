<script>
import { GlAvatar } from '@gitlab/ui';
import SafeHtml from '~/vue_shared/directives/safe_html';
import NoteHeader from '~/notes/components/note_header.vue';
import TimelineEntryItem from '~/vue_shared/components/notes/timeline_entry_item.vue';

export default {
  name: 'WorkItemNoteReplying',
  components: {
    TimelineEntryItem,
    GlAvatar,
    NoteHeader,
  },
  directives: {
    SafeHtml,
  },
  safeHtmlConfig: {
    ADD_TAGS: ['use', 'gl-emoji', 'copy-code'],
  },
  constantOptions: {
    avatarUrl: window.gon.current_user_avatar_url,
  },
  props: {
    body: {
      type: String,
      required: false,
      default: '',
    },
  },
  computed: {
    author() {
      return {
        avatarUrl: window.gon.current_user_avatar_url,
        id: window.gon.current_user_id,
        name: window.gon.current_user_fullname,
        username: window.gon.current_username,
      };
    },
  },
};
</script>

<template>
  <timeline-entry-item class="note note-wrapper note-comment gl-p-4 being-posted">
    <div class="timeline-avatar gl-float-left">
      <gl-avatar :src="$options.constantOptions.avatarUrl" :size="32" class="gl-mr-3" />
    </div>
    <div class="note-header">
      <note-header :author="author" />
    </div>
    <div ref="note-body" v-safe-html:[$options.safeHtmlConfig]="body" class="note-body"></div>
  </timeline-entry-item>
</template>
