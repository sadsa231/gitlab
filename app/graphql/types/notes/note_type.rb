# frozen_string_literal: true

module Types
  module Notes
    class NoteType < BaseObject
      graphql_name 'Note'

      authorize :read_note

      expose_permissions Types::PermissionTypes::Note

      implements(Types::ResolvableInterface)

      field :id, ::Types::GlobalIDType[::Note],
        null: false,
        description: 'ID of the note.'

      field :project, Types::ProjectType,
        null: true,
        description: 'Project associated with the note.'

      field :author, Types::UserType,
        null: false,
        description: 'User who wrote this note.'

      field :system, GraphQL::Types::Boolean,
        null: false,
        description: 'Indicates whether this note was created by the system or by a user.'
      field :system_note_icon_name,
        GraphQL::Types::String,
        null: true,
        description: 'Name of the icon corresponding to a system note.'

      field :body, GraphQL::Types::String,
        null: false,
        method: :note,
        description: 'Content of the note.'

      field :confidential, GraphQL::Types::Boolean,
        null: true,
        description: 'Indicates if this note is confidential.',
        method: :confidential?,
        deprecated: {
          reason: :renamed,
          replacement: 'internal',
          milestone: '15.5'
        }

      field :internal, GraphQL::Types::Boolean,
        null: true,
        description: 'Indicates if this note is internal.',
        method: :confidential?

      field :created_at, Types::TimeType,
        null: false,
        description: 'Timestamp of the note creation.'
      field :discussion, Types::Notes::DiscussionType,
        null: true,
        description: 'Discussion this note is a part of.'
      field :position, Types::Notes::DiffPositionType,
        null: true,
        description: 'Position of this note on a diff.'
      field :updated_at, Types::TimeType,
        null: false,
        description: "Timestamp of the note's last activity."
      field :url, GraphQL::Types::String,
        null: true,
        description: 'URL to view this Note in the Web UI.'

      field :last_edited_at, Types::TimeType,
        null: true,
        description: 'Timestamp when note was last edited.'
      field :last_edited_by, Types::UserType,
        null: true,
        description: 'User who last edited the note.'

      field :system_note_metadata, Types::Notes::SystemNoteMetadataType,
        null: true,
        description: 'Metadata for the given note if it is a system note.'

      markdown_field :body_html, null: true, method: :note

      def url
        ::Gitlab::UrlBuilder.build(object)
      end

      def system_note_icon_name
        SystemNoteHelper.system_note_icon_name(object) if object.system?
      end

      def project
        Gitlab::Graphql::Loaders::BatchModelLoader.new(Project, object.project_id).find
      end

      def author
        Gitlab::Graphql::Loaders::BatchModelLoader.new(User, object.author_id).find
      end

      # We now support also SyntheticNote notes as a NoteType, but SyntheticNote does not have a real note ID,
      # as SyntheticNote is generated dynamically from a ResourceEvent instance.
      def id
        return super unless object.is_a?(SyntheticNote)

        ::Gitlab::GlobalId.build(object, model_name: object.class.to_s, id: object.discussion_id)
      end
    end
  end
end
