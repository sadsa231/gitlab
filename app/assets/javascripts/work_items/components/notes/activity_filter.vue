<script>
import { GlDropdown, GlDropdownItem } from '@gitlab/ui';
import { __ } from '~/locale';
import Tracking from '~/tracking';
import LocalStorageSync from '~/vue_shared/components/local_storage_sync.vue';
import { ASC, DESC } from '~/notes/constants';
import { TRACKING_CATEGORY_SHOW, WORK_ITEM_NOTES_SORT_ORDER_KEY } from '~/work_items/constants';

const SORT_OPTIONS = [
  { key: DESC, text: __('Newest first'), dataid: 'js-newest-first' },
  { key: ASC, text: __('Oldest first'), dataid: 'js-oldest-first' },
];

export default {
  SORT_OPTIONS,
  components: {
    GlDropdown,
    GlDropdownItem,
    LocalStorageSync,
  },
  mixins: [Tracking.mixin()],
  props: {
    sortOrder: {
      type: String,
      default: ASC,
      required: false,
    },
    loading: {
      type: Boolean,
      default: false,
      required: false,
    },
    workItemType: {
      type: String,
      required: true,
    },
  },
  data() {
    return {
      persistSortOrder: true,
    };
  },
  computed: {
    tracking() {
      return {
        category: TRACKING_CATEGORY_SHOW,
        label: 'item_track_notes_sorting',
        property: `type_${this.workItemType}`,
      };
    },
    selectedSortOption() {
      const isSortOptionValid = this.sortOrder === ASC || this.sortOrder === DESC;
      return isSortOptionValid ? SORT_OPTIONS.find(({ key }) => this.sortOrder === key) : ASC;
    },
    getDropdownSelectedText() {
      return this.selectedSortOption.text;
    },
  },
  methods: {
    setDiscussionSortDirection(direction) {
      this.$emit('updateSavedSortOrder', direction);
    },
    fetchSortedDiscussions(direction) {
      if (this.isSortDropdownItemActive(direction)) {
        return;
      }
      this.track('notes_sort_order_changed');
      this.$emit('changeSortOrder', direction);
    },
    isSortDropdownItemActive(sortDir) {
      return sortDir === this.sortOrder;
    },
  },
  WORK_ITEM_NOTES_SORT_ORDER_KEY,
};
</script>

<template>
  <div
    id="discussion-preferences"
    data-testid="discussion-preferences"
    class="gl-display-inline-block gl-vertical-align-bottom gl-w-full gl-sm-w-auto"
  >
    <local-storage-sync
      :value="sortOrder"
      :storage-key="$options.WORK_ITEM_NOTES_SORT_ORDER_KEY"
      :persist="persistSortOrder"
      as-string
      @input="setDiscussionSortDirection"
    />
    <gl-dropdown
      :id="`discussion-preferences-dropdown-${workItemType}`"
      class="gl-xs-w-full"
      size="small"
      :text="getDropdownSelectedText"
      :disabled="loading"
      right
    >
      <div id="discussion-sort">
        <gl-dropdown-item
          v-for="{ text, key, dataid } in $options.SORT_OPTIONS"
          :key="text"
          :data-testid="dataid"
          is-check-item
          :is-checked="isSortDropdownItemActive(key)"
          @click="fetchSortedDiscussions(key)"
        >
          {{ text }}
        </gl-dropdown-item>
      </div>
    </gl-dropdown>
  </div>
</template>
