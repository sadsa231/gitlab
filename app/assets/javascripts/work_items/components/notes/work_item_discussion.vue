<script>
import { GlAvatarLink, GlAvatar } from '@gitlab/ui';
import { ASC } from '~/notes/constants';
import TimelineEntryItem from '~/vue_shared/components/notes/timeline_entry_item.vue';
import DiscussionNotesRepliesWrapper from '~/notes/components/discussion_notes_replies_wrapper.vue';
import ToggleRepliesWidget from '~/notes/components/toggle_replies_widget.vue';
import WorkItemNote from '~/work_items/components/notes/work_item_note.vue';
import WorkItemNoteReplying from '~/work_items/components/notes/work_item_note_replying.vue';
import WorkItemCommentForm from '../work_item_comment_form.vue';

export default {
  components: {
    TimelineEntryItem,
    GlAvatarLink,
    GlAvatar,
    WorkItemNote,
    WorkItemCommentForm,
    ToggleRepliesWidget,
    DiscussionNotesRepliesWrapper,
    WorkItemNoteReplying,
  },
  props: {
    workItemId: {
      type: String,
      required: true,
    },
    queryVariables: {
      type: Object,
      required: true,
    },
    fullPath: {
      type: String,
      required: true,
    },
    workItemType: {
      type: String,
      required: true,
    },
    fetchByIid: {
      type: Boolean,
      required: false,
      default: false,
    },
    discussion: {
      type: Array,
      required: true,
    },
    sortOrder: {
      type: String,
      default: ASC,
      required: false,
    },
  },
  data() {
    return {
      isExpanded: false,
      autofocus: false,
      isReplying: false,
      replyingText: '',
    };
  },
  computed: {
    note() {
      return this.discussion[0];
    },
    author() {
      return this.note.author;
    },
    noteAnchorId() {
      return `note_${this.note.id}`;
    },
    hasReplies() {
      return this.replies?.length;
    },
    replies() {
      if (this.discussion?.length > 1) {
        return this.discussion.slice(1);
      }
      return null;
    },
    discussionId() {
      return this.discussion[0]?.discussion?.id || '';
    },
  },
  methods: {
    showReplyForm() {
      this.isExpanded = true;
      this.autofocus = true;
    },
    hideReplyForm() {
      this.isExpanded = this.hasReplies;
      this.autofocus = false;
    },
    toggleDiscussion() {
      this.isExpanded = !this.isExpanded;
      this.autofocus = this.isExpanded;
    },
    threadKey(note) {
      /* eslint-disable @gitlab/require-i18n-strings */
      return `${note.id}-thread`;
    },
    onReplied() {
      this.isExpanded = true;
      this.isReplying = false;
      this.replyingText = '';
    },
    onReplying(commentText) {
      this.isReplying = true;
      this.replyingText = commentText;
    },
  },
};
</script>

<template>
  <timeline-entry-item
    :id="noteAnchorId"
    :class="{ 'internal-note': note.internal }"
    :data-note-id="note.id"
    class="note note-wrapper note-comment gl-px-0"
  >
    <div class="timeline-avatar gl-float-left">
      <gl-avatar-link :href="author.webUrl">
        <gl-avatar
          :src="author.avatarUrl"
          :entity-name="author.username"
          :alt="author.name"
          :size="32"
        />
      </gl-avatar-link>
    </div>

    <div class="timeline-content">
      <div class="discussion-body">
        <div class="discussion-wrapper">
          <div class="discussion-notes">
            <ul class="notes">
              <work-item-note
                :is-first-note="true"
                :note="note"
                :discussion-id="discussionId"
                @startReplying="showReplyForm"
              />
              <discussion-notes-replies-wrapper>
                <toggle-replies-widget
                  v-if="hasReplies"
                  :collapsed="!isExpanded"
                  :replies="replies"
                  @toggle="toggleDiscussion({ discussionId })"
                />
                <template v-if="isExpanded">
                  <template v-for="reply in replies">
                    <work-item-note
                      :key="threadKey(reply)"
                      discussion-id="discussionId"
                      :note="reply"
                      @startReplying="showReplyForm"
                    />
                  </template>
                  <work-item-note-replying v-if="isReplying" :body="replyingText" />
                  <work-item-comment-form
                    :autofocus="autofocus"
                    :query-variables="queryVariables"
                    :full-path="fullPath"
                    :work-item-id="workItemId"
                    :fetch-by-iid="fetchByIid"
                    :discussion-id="discussionId"
                    :work-item-type="workItemType"
                    :sort-order="sortOrder"
                    :add-padding="true"
                    @cancelEditing="hideReplyForm"
                    @replied="onReplied"
                    @replying="onReplying"
                    @error="$emit('error', $event)"
                  />
                </template>
              </discussion-notes-replies-wrapper>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </timeline-entry-item>
</template>
