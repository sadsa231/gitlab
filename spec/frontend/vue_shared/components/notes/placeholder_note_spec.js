import { shallowMount } from '@vue/test-utils';
import Vue from 'vue';
import Vuex from 'vuex';
import IssuePlaceholderNote from '~/vue_shared/components/notes/placeholder_note.vue';
import { userDataMock } from 'jest/notes/mock_data';

Vue.use(Vuex);

const getters = {
  getUserData: () => userDataMock,
};

describe('Issue placeholder note component', () => {
  let wrapper;

  const findNote = () => wrapper.findComponent({ ref: 'note' });

  const createComponent = (isIndividual = false, propsData = {}) => {
    wrapper = shallowMount(IssuePlaceholderNote, {
      store: new Vuex.Store({
        getters,
      }),
      propsData: {
        note: {
          body: 'Foo',
          individual_note: isIndividual,
        },
        ...propsData,
      },
    });
  };

  afterEach(() => {
    wrapper.destroy();
    wrapper = null;
  });

  it('matches snapshot', () => {
    createComponent();

    expect(wrapper.element).toMatchSnapshot();
  });

  it('does not add "discussion" class to individual notes', () => {
    createComponent(true);

    expect(findNote().classes()).not.toContain('discussion');
  });

  it('adds "discussion" class to non-individual notes', () => {
    createComponent();

    expect(findNote().classes()).toContain('discussion');
  });
});
