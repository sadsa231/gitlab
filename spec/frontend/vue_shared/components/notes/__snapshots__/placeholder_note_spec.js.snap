// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Issue placeholder note component matches snapshot 1`] = `
<timeline-entry-item-stub
  class="note note-wrapper note-comment being-posted fade-in-half"
>
  <div
    class="timeline-avatar gl-float-left"
  >
    <gl-avatar-link-stub
      href="/root"
    >
      <gl-avatar-stub
        alt="Root"
        entityid="0"
        entityname="root"
        shape="circle"
        size="32"
        src="mock_path"
      />
    </gl-avatar-link-stub>
  </div>
   
  <div
    class="timeline-content discussion"
  >
    <div
      class="note-header"
    >
      <div
        class="note-header-info"
      >
        <a
          href="/root"
        >
          <span
            class="d-none d-sm-inline-block bold"
          >
            Root
          </span>
           
          <span
            class="note-headline-light"
          >
            @root
          </span>
        </a>
      </div>
    </div>
     
    <div
      class="timeline-discussion-body"
    >
      <div
        class="note-body"
      >
        <div
          class="note-text md"
        >
          <p>
            Foo
          </p>
          

        </div>
      </div>
    </div>
  </div>
</timeline-entry-item-stub>
`;
